### PR TITLE
feat(engine): improve type hints

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
This PR aims to improve and add to the currently existing type hints in the `engine` submodule.

### CODE REVIEW

File: nerfstudio/engine/trainer.py [hunk 1]
```diff
@@ -104,39 +100,39 @@ class Trainer:
...
     callbacks: List[TrainingCallback]
 
-    def __init__(self, config: TrainerConfig, local_rank: int = 0, world_size: int = 1):
+    def __init__(self, config: TrainerConfig, local_rank: int = 0, world_size: int = 1) -> None:
         self.config = config
         self.local_rank = local_rank
         self.world_size = world_size
-        self.device = "cpu" if world_size == 0 else f"cuda:{local_rank}"
-        self.mixed_precision = self.config.mixed_precision
+        self.device: str = "cpu" if world_size == 0 else f"cuda:{local_rank}"
```
Comment (tancik) [inline nerfstudio/engine/trainer.py:115]: Maybe this should be `Union[torch.device, str]`

**Review by tancik — COMMENTED** [review]
Thanks for helping to clean up this code!

File: nerfstudio/engine/trainer.py (hunk 1 above)
Comment (SauravMaheshkar) [reply nerfstudio/engine/trainer.py]: Addressed in [71cadf2](https://github.com/nerfstudio-project/nerfstudio/pull/1509/commits/71cadf2d0ebd7815de1590340174268ff954e349)

**Review by tancik — APPROVED** [review]
Lgtm! Thanks!

**Pull request merged**